### ISSUE TITLE
[c10d] Use torch::autograd::Variable with no_grad

### ISSUE THREAD

**Issue by pietern** [pytorch/pytorch#19145]
We currently use `at::Tensor` directly and don't worry about the distinction between Variable and Tensor in c10d. This came up when working on adding support for reduction of sparse tensors, where Variable instances that come in through pybind are mixed with temporary Tensor instances created on the c10d side.

One solution here is to use `torch::autograd::Variable` (later to be fully merged into `at::Tensor`) throughout. This means adding a dependency on `torch/csrc/autograd` in c10d and changing all temporaries to be created with factory functions in the `torch` namespace.

A temporary workaround is to unbox and rebox Variable instances at the Python/C++ boundary.

**Comment by pietern** [comment]
Discussed this with @gchanan.

**Comment by pietern** [comment]
@gchanan After rebasing my diffs it looks like this is no longer an issue. I might have missed that the Tensor/Variable merge has landed. Can you confirm that any new tensor through `at::empty` (for example) can act like a Variable now?

**Comment by gchanan** [comment]
CC @yf225.

I don't believe that's correct, although I guess it depends on what you mean by "act like a Variable."  What's your definition?

**Comment by yf225** [comment]
@pietern After #17072, we don't unpack Variables when we pass them into ATen ops (instead we use `at::AutoNonVariableTypeMode` so that the ATen ops will use the Variables as if they are Tensors). The tensors returned from ATen ops can still be either Tensors or Variables, and we need to check whether they are Tensors to decide whether to rebox them or not.

**Comment by gchanan** [comment]
@yf225 @pietern ya, this pretty tricky and brittle...

**Comment by ezyang** [comment]
Now that Tensor-Variable is merged, there are no more problems here.
